crates/component/src/wrappers/http.rs

```diff
@@ -0,0 +1,353 @@
+//! This module provides utilities for writing HTTP servers and clients using the WASI HTTP API.
+//!
+//! It's inspired by the WASI 0.3 proposal for <https://github.com/WebAssembly/wasi-http> and will
+//! be supported until the release of wasi:http@0.3.0. After that, this module will be deprecated.
+// Required until <https://github.com/seanmonstar/reqwest/pull/2453> merges
+use reqwest_wasmcloud as reqwest;
+
+use std::{
+    io::{Read, Write},
+    str::FromStr,
+};
+
+use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
+use wasi::{
+    http::{
+        outgoing_handler::ErrorCode,
+        types::{
+            Fields, IncomingBody, InputStream, Method, OutgoingBody, OutgoingResponse,
+            ResponseOutparam, Scheme,
+        },
+    },
+    io::streams::StreamError,
+};
+
+const HTTP_SCHEME: &str = "http";
+const HTTPS_SCHEME: &str = "https";
+
+/// Trait for implementing an HTTP server WebAssembly component that receives a
+/// [`Request`] and returns a [`ResponseBuilder`].
+pub trait HttpServer {
+    fn handle(request: Request) -> Result<ResponseBuilder, ErrorCode>;
+}
+
+pub struct Request {
+    inner: wasi::http::types::IncomingRequest,
+}
+
+impl TryFrom<wasi::http::types::IncomingRequest> for Request {
+    type Error = ErrorCode;
+
+    fn try_from(inner: wasi::http::types::IncomingRequest) -> Result<Self, Self::Error> {
+        Ok(Self { inner })
+    }
+}
+
+impl Request {
+    pub fn method(&self) -> Method {
+        self.inner.method()
+    }
+
+    pub fn headers(&self) -> Fields {
+        self.inner.headers()
+    }
+
+    pub fn scheme(&self) -> Option<Scheme> {
+        self.inner.scheme()
+    }
+
+    pub fn authority(&self) -> Option<String> {
+        self.inner.authority()
+    }
+
+    pub fn path_with_query(&self) -> Option<String> {
+        self.inner.path_with_query()
+    }
+
+    /// Read the entire body of the [`Request`] into a buffer and return it.
+    /// This consumes the request and finishes the body.
+    ///
+    /// For very large bodies, consider using [`Request::into_body_stream`] instead.
+    pub fn into_body(self) -> Result<Vec<u8>, ErrorCode> {
+        let (mut input_stream, incoming_body) = self.consume_request()?;
+        let mut buf = vec![];
+        input_stream
+            .read_to_end(&mut buf)
+            .map_err(|_| error_code("failed to read incoming body"))?;
+        drop(input_stream);
+        IncomingBody::finish(incoming_body);
+        Ok(buf)
+    }
+
+    /// Converts the request into the inner [`InputStream`] and [`IncomingBody`]
+    ///
+    /// These values have a tightly coupled lifecycle and the stream must be dropped
+    /// before the body is finished.
+    ///
+    /// Use this in tandem with [`ResponseBuilder::ok_stream`] to return a response with a body stream.
```

Comment:
doesn't look like `ok_stream` exists

Reply:
Yeah this got overwritten over a refactor or two, fixing